compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp

```diff
@@ -65,26 +66,225 @@ static FailureOr<AffineMap> getComposedAffineMap(Attribute attr) {
 EncodingAttr EncodingAttr::get(MLIRContext *ctx, int64_t operandIndex,
                                EncodingOpType opType, ArrayRef<Type> elemTypes,
                                ArrayRef<AffineMap> maps,
-                               ArrayRef<int64_t> roundDimsTo,
+                               ArrayRef<int64_t> iterationSizes,
                                ArrayRef<Attribute> layouts) {
   Builder b(ctx);
   auto opTypeAttr = EncodingOpTypeAttr::get(ctx, opType);
   auto mapsAttr = maps.empty() ? ArrayAttr() : b.getAffineMapArrayAttr(maps);
-  auto roundDimsToAttr = roundDimsTo.empty()
-                             ? DenseI64ArrayAttr()
-                             : b.getDenseI64ArrayAttr(roundDimsTo);
+  auto iterationSizesAttr =
+      iterationSizes.empty() ? ArrayAttr() : b.getI64ArrayAttr(iterationSizes);
   auto layoutsAttr = layouts.empty() ? ArrayAttr() : b.getArrayAttr(layouts);
   return get(ctx, b.getIndexAttr(operandIndex), opTypeAttr,
-             b.getTypeArrayAttr(elemTypes), mapsAttr, roundDimsToAttr,
+             b.getTypeArrayAttr(elemTypes), mapsAttr, iterationSizesAttr,
              layoutsAttr);
 }
 
+/// Parse a comma-separated list of key-value pairs with a specified
+/// delimiter. This performs similar parsing as the the assembly format
+/// `struct` directive parser with a specified delimiter. The variables are
+/// printed in the order they are specified in the argument list but can be
+/// parsed in any order.
+///
+/// Example:
+/// <
+///   foo = something_parsed_by_a_custom_parser,
+///   bar = something_parsed_by_a_different_custom_parser,
+///   ...
+/// >
+static ParseResult
+parseStruct(AsmParser &p, AsmParser::Delimiter delimiter,
+            ArrayRef<StringRef> keywords,
+            ArrayRef<llvm::function_ref<ParseResult()>> parseFuncs) {
+  assert(keywords.size() == parseFuncs.size());
+  auto keyError = [&]() -> ParseResult {
+    InFlightDiagnostic parseError =
+        p.emitError(p.getCurrentLocation(), "expected one of: ");
+    llvm::interleaveComma(keywords, parseError, [&](StringRef kw) {
+      parseError << '`' << kw << '`';
+    });
+    return parseError;
+  };
+  SmallVector<bool> seen(keywords.size(), false);
+  DenseMap<StringRef, size_t> keywordToIndex;
+  for (auto &&[idx, keyword] : llvm::enumerate(keywords))
+    keywordToIndex[keyword] = idx;
```

Comment:
IREE follows mixed Google/MLIR style, mostly MLIR style. In IREE, we wrap simple single-statements with braces.

Why do we need rvalue reference (i.e., `&&`)? Is it because the `keywords` has `ArrayRef<StringRef>` type?

```suggestion
  for (auto [&&idx, keyword] : llvm::enumerate(keywords)) {
    keywordToIndex[keyword] = idx;
  }
```

Reply:
I don't think `auto &&[idx, keyword]` does anything wrong here, does it?

Reply:
No, it does not. I was trying to clarify the code.

Reply:
`auto &&` is my go-to standard in ranges as I don't think you can go wrong with it, except for clarity maybe.  Updated it to `auto [idx, keyword]` here for the latter if that's the style guide.